### DISCUSSION TITLE
Water and Steam Dominated Hydrothermal System

### DISCUSSION THREAD

**Discussion by geoyanzhan3** [idaholab/moose#25493] [Q&A General]
Dear all, 

Our group tried to implement the two-phase (phase change between liquid and water vapor) flow in a porous media using the same governing equations published in ([Faust and Mercer, 1979](http://doi.wiley.com/10.1029/WR015i001p00023)). 

We tried to use COMSOL, but COMSOL doesn't use enthalpy as the variable for energy balance. We wrote our own PDE in COMSOL, but it was not very stable, and super slow. 

I wonder if MOOSE can be used to solve those equations as attached. I guess MOOSE already has the [water-steam table](https://mooseframework.inl.gov/modules/porous_flow/water_vapor.html) for getting the parameters dependent on $p, h$. 

Thank you so much! 
Best, 
Yan

### The governing equations for the problem. 

$$ \frac{\partial (\phi \rho)}{\partial t} - \nabla \cdot [ \frac{K k_{rs} \rho_s} {\mu_s} \cdot (\nabla p - \rho_s g )] - \nabla \cdot [ \frac{K k_{rw} \rho_w} {\mu_w} \cdot (\nabla p - \rho_w g )] = 0 $$

$$ \frac{\partial}{\partial t}[\phi \rho h + (1 - \phi) \rho_r h_r] - \nabla \cdot [ \frac{K k_{rs} \rho_s h_s} {\mu_s} \cdot (\nabla p - \rho_s g )] - \nabla \cdot [ \frac{K k_{rw} \rho_w h_w} {\mu_w} \cdot (\nabla p - \rho_w g )] -  \nabla \cdot [K_m (\frac{\partial T}{\partial p})_h \nabla p + K_m (\frac{\partial T}{\partial h})_p \nabla h]= 0 $$

Variable: 
- $p$ = Pressure 
- $h$ = Enthalpy

Parameters dependent on $p, h$ (can be acquired by the water-steam table):
- T = temperature 
- $h_r = C_p^r T$ = rock enthalpy
- $k_{rw}$ = relative permeability of water
- $k_{rs}$ = relative permeability of steam
- $\rho_w$ = density of water
- $\rho_s$ = density of steam
- $h_w$ = enthalpy of water
- $h_s$ = enthalpy of steam
- $\mu_w$ = viscosity of water
- $\mu_s$ = viscosity of steam

Parameters independent on $p, h$
- $\phi$ = Porosity 
- $K$ = Intrinsic permeability
- $K_m$ = medium thermal conductivity
- $\rho_r$ = rock density
- $C_p^r$ = rock specific heat capacity

**Comment by GiudGiud** [comment]
@cpgr @rpodgorney

**Comment by cpgr** [comment]
We do have some limited pressure - enthalpy capability for water and steam in porous media - see an example https://mooseframework.inl.gov/modules/porous_flow/water_vapor.html

**Reply by geoyanzhan3** [reply]
Thank you! I have already read that page. Do you know how to set the initial water saturation if the initial fluid state is two-phase? For example, I want to benchmark the code with the Problem 2B in the DOE code comparison project. As attached 
![IMG_0043](https://github.com/idaholab/moose/assets/35808906/47266518-7b73-4291-b6dc-fc5204fd2d58)
![IMG_0044](https://github.com/idaholab/moose/assets/35808906/5601bf35-f91a-4ec3-ad06-93dda17854da)

**Reply by geoyanzhan3** [reply]
In addition, I benchmarked the problem in COMSOL by using temperature as the variable for the energy. And define a transition around the boiling temperature of the water to make the water or stream quality as a function of the temperature. Then, calculate the saturation to get other parameters. But this method doesn’t work for other problems. Therefore, I want to try the original formulation by Faust et al. However, comsol seems does not like the enthalpy setting.

**Reply by cpgr** [reply]
Oh, I just checked and we have an IC for the more complicated models that computes everything given an initial saturation, but not for this one. It wouldn't be hard to make, but you could just work out the initial enthalpy by hand for this case just rearranging Eq 3 of the linked page.

**Reply by geoyanzhan3** [reply]
Thank you @cpgr !

I have created an input file based on [modules/porous_flow/test/tests/fluidstate/water_vapor.i](https://mooseframework.inl.gov/modules/porous_flow/water_vapor.html#e12b5988-9cb7-4e52-a159-eaf20aaeb808) to solve the [problem 2B](https://github.com/idaholab/moose/discussions/25493#discussioncomment-7041023) that I mentioned. 

I changed the parameters and the initial value to make the initial sw~0.65. I added a mass outflow boundary to the left:
```
[BCs]
  [drainage]
    type = PorousFlowSink
    variable = pliq
    boundary = left
    flux_function = -0.053
  []
[]
```

I wonder how to set the capillary pressure as zero. And can I make the timestep increase exponentially? 

I am attaching the input file. Thank you very much! 

[water_vapor_problem2b.i.zip](https://github.com/idaholab/moose/files/12656571/water_vapor_problem2b.i.zip)

**Reply by geoyanzhan3** [reply]
I modified the executioner to use the iteration adaptive stepper. 

But the model doesn't converge after several time steps.  I appreciate any help in figuring out why it happens

I am also attaching the updated input file. Thank you!
[water_vapor_problem2b.i.zip](https://github.com/idaholab/moose/files/12679424/water_vapor_problem2b.i.zip)



```
[Executioner]
  type = Transient
  solve_type = NEWTON
  dtmin = 0.01
  dtmax = 100
  end_time = 1e5
  nl_abs_tol = 1e-5
    [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 5
    dt = 1.0
  []
[]
```

**Reply by GiudGiud** [reply]
these instructions may be able to help in the mean time
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by cpgr** [reply]
It's almost certainly your BC that is causing the issue. At the moment, it is taking water only, no steam, and only mass, not heat.

I assume from the benchmark problem that it should take both water and steam (mass and heat), at a ration given by the relatively perms?

**Reply by cpgr** [reply]
PS - a zero capillary pressure can be used pretty easily:
```
type = PorousFlowCapillaryPressureConst
pc = 0
```

**Reply by geoyanzhan3** [reply]
Thank you very much! Do you know which BC module can I use? Or can I keep use "PorousFlowSink" but modified the parameters?

```
[BCs]
  [drainage]
    type = PorousFlowSink
    variable = pliq
    boundary = left
    flux_function = -0.053158
  []
[]
```
